hal/network/ncp_client/sara/sara_ncp_client.cpp

```diff
@@ -1044,28 +1053,64 @@ int SaraNcpClient::selectSimCard(ModemState& state) {
-        // Set UMNOPROF = SIM_SELECT
-        auto resp = parser_.sendCommand("AT+UMNOPROF?");
-        reset = false;
-        int umnoprof = static_cast<int>(UbloxSaraUmnoprof::NONE);
-        auto r = CHECK_PARSER(resp.scanf("+UMNOPROF: %d", &umnoprof));
-        CHECK_PARSER_OK(resp.readResult());
-        if (r == 1 && static_cast<UbloxSaraUmnoprof>(umnoprof) == UbloxSaraUmnoprof::SW_DEFAULT) {
-            // Disconnect before making changes to the UMNOPROF
-            r = CHECK_PARSER(parser_.execCommand("AT+COPS=2,2"));
-            if (r == AtResponse::OK) {
+        int resetCount = 0;
+        int isTwilio = 0;
+        // Check if we are using a Twilio Super SIM based on ICCID
+        char buf[32] = {};
+        auto lenCcid = getIccidImpl(buf, sizeof(buf));
+        CellularNetworkConfig netConfig;
+        CHECK_TRUE(lenCcid > 5, SYSTEM_ERROR_AT_NOT_OK);
+        netConfig = networkConfigForIccid(buf, lenCcid);
+        if (!strcmp(netConfig.apn(), "super")) { // if Twilio Super SIM
+            isTwilio = 1;
+        }
+        do {
+            // Set UMNOPROF = SIM_SELECT
+            auto resp = parser_.sendCommand("AT+UMNOPROF?");
+            reset = false;
+            int umnoprof = static_cast<int>(UbloxSaraUmnoprof::NONE);
+            auto r = CHECK_PARSER(resp.scanf("+UMNOPROF: %d", &umnoprof));
+            CHECK_PARSER_OK(resp.readResult());
+            if (r == 1 && static_cast<UbloxSaraUmnoprof>(umnoprof) == UbloxSaraUmnoprof::SW_DEFAULT) {
+                // Disconnect before making changes to the UMNOPROF
+                auto respCfun = parser_.sendCommand(UBLOX_CFUN_TIMEOUT, "AT+CFUN?");
+                int cfunVal = -1;
+                auto retCfun = CHECK_PARSER(respCfun.scanf("+CFUN: %d", &cfunVal));
+                CHECK_PARSER_OK(respCfun.readResult());
+                if (retCfun == 1 && cfunVal != 0) {
+                    CHECK_PARSER_OK(parser_.execCommand(UBLOX_CFUN_TIMEOUT, "AT+CFUN=0"));
+                }
+
                 // This is a persistent setting
-                auto respUmno = parser_.sendCommand(1000, "AT+UMNOPROF=%d", static_cast<int>(UbloxSaraUmnoprof::SIM_SELECT));
-                respUmno.readResult();
+                int umnoprof = static_cast<int>(UbloxSaraUmnoprof::SIM_SELECT);
+                if (isTwilio) { // if Twilio Super SIM
+                    umnoprof = static_cast<int>(UbloxSaraUmnoprof::STANDARD_EUROPE);
+                }
+                parser_.execCommand(1000, "AT+UMNOPROF=%d", umnoprof);
                 // Not checking for error since we will reset either way
                 reset = true;
+            } else if (r == 1 && static_cast<UbloxSaraUmnoprof>(umnoprof) == UbloxSaraUmnoprof::STANDARD_EUROPE) {
+                auto respBand = parser_.sendCommand(UBLOX_UBANDMASK_TIMEOUT, "AT+UBANDMASK?");
+                uint64_t ubandCatm1 = 0;
+                auto retBand = CHECK_PARSER(respBand.scanf("+UBANDMASK: 0,%llu", &ubandCatm1));
```

Comment:
This will not work. We use `newlib-nano` variant of libc, which does not support 64-bit integers in `scanf` and `printf` type of functions. `strtoull` should work though.

Reply:
Good catch, I'll refactor this one.